### PR TITLE
feat: Window func

### LINKED ISSUE #4653: Feature:  Window Functions Support
**Summary**

Description for this feature.

1. Syntax: https://dev.mysql.com/doc/refman/8.0/en/window-functions-usage.html
2. Paper: http://www.vldb.org/pvldb/vol8/p1058-leis.pdf

### PR DESCRIPTION
**Pull request by doki23**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary
impl it partially including
1. aggr function
2. const rows frame
3. const range frame
4. only works in the old scheduler framework now

## Changelog

- New Feature
aggregate window function

## Related Issues

Closes #4653

### CODE REVIEW

**Comment by Xuanwo** [comment]
@mergify update

File: common/functions/src/window/mod.rs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+mod window_frame;
```
Comment (BohuTANG) [inline common/functions/src/window/mod.rs:1]: Hmm, need a license  header to make the lint happy :)

**Comment by doki23** [comment]
Sorry for the late update, because I'm too busy recently. 
It's ready for review now.

File: common/functions/src/window/window_frame.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+use std::hash::Hash;
+use std::hash::Hasher;
+
+use common_exception::ErrorCode;
+use sqlparser::ast;
+
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Hash, serde::Serialize, serde::Deserialize,
+)]
+pub struct WindowFrame {
```
Comment (doki23) [inline common/functions/src/window/window_frame.rs:26]: Code in this file is mainly copied from another open source project -- apache/arrow-datafusion. And I'm not sure whether it's proper or if I should add some description on the top of it.

Comment (Xuanwo) [reply common/functions/src/window/window_frame.rs:26]: Please attach their license header in this file with a clear statement, for example:

https://github.com/apache/arrow-datafusion/blob/master/datafusion/core/src/prelude.rs#L1-L16

And add descriptions include:

- Original Project
- Link to commits

Comment (doki23) [reply common/functions/src/window/window_frame.rs:26]: > I think it's ok. Do we have stateless tests for this pr? Like: https://github.com/datafuselabs/databend/blob/49252766894f3bb5f5d4767037142cb1b10d92bc/tests/suites/0_stateless/03_dml/03_0003_select_group_by.sql

Thanks for the remind and I'll add some stateless tests then.

**Comment by sundy-li** [comment]
Impressive pr, I'll take time to review it.

**Review by zhang2014 — APPROVED** [review]
👍  Look good to me. BTW: need implement window function in new processor.

**Comment by BohuTANG** [comment]
@mergify update

**Review by BohuTANG — APPROVED** [review]
👍

**Comment by doki23** [comment]
> 👍 Look good to me. BTW: need implement window function in new processor.

Yes, I'll implement it in new processor in a follow-up issue and also make some optimizations about performance.

File: query/src/pipelines/transforms/transform_window_func.rs [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+        let mut arguments: Vec<ColumnWithField> = Vec::with_capacity(args.len());
+        let mut arg_fields = Vec::with_capacity(args.len());
+        for arg in args {
+            let arg_field: DataField = arg.to_data_field(&self.input_schema)?;
+            arg_fields.push(arg_field.clone());
+            let arg_column = block.try_column_by_name(arg_field.name()).unwrap();
+            arguments.push(ColumnWithField::new(Arc::clone(arg_column), arg_field));
+        }
+
+        let function = if !AggregateFunctionFactory::instance().check(op) {
```
Comment (sundy-li) [inline query/src/pipelines/transforms/transform_window_func.rs:152]: we can have a function field in this transformer.

Comment (sundy-li) [inline query/src/pipelines/transforms/transform_window_func.rs:152]: we can have a function field in this transformer.

File: query/src/pipelines/transforms/transform_window_func.rs [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+                let frame_start = frame.start;
+                let frame_end = frame.end;
+                let state = segment_tree_state.query(frame_start, frame_end);
+                let mut builder = function.return_type().unwrap().create_mutable(1);
+                function.merge_result(state, builder.as_mut()).unwrap();
+                builder.to_column()
+            })
+            .collect::<Vec<_>>();
+
+        let window_col = Series::concat(&window_col_per_tuple).unwrap();
```
Comment (sundy-li) [inline query/src/pipelines/transforms/transform_window_func.rs:182]: unwrap could be replaced by  Result.

File: query/src/pipelines/transforms/transform_window_func.rs [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+        let rows = arguments[0].len();
+        let state_per_tuple = (0..rows)
+            .map(|i| {
+                arguments
+                    .iter()
+                    .map(|c| c.slice(i, 1))
+                    .collect::<Vec<ColumnRef>>()
+            })
+            .map(|args| {
+                let place = arena.alloc_layout(func.state_layout());
```
Comment (sundy-li) [inline query/src/pipelines/transforms/transform_window_func.rs:457]: we should call drop_states to destroy the inner allocation  as `PartialAggregator` did.

**Review by sundy-li — COMMENTED** [review]
Left some comments, these can be improved in another pr.

**Comment by BohuTANG** [comment]
Ping @soyeric128 for documentation.

**Comment by soyeric128** [comment]
Hi @BohuTANG @doki23 

Could you pls help confirm?
- Do we support ALL of these aggregate functions for window functions: https://databend.rs/doc/reference/functions/aggregate-functions
- Do we now support any of non-aggr window functions?

Thanks.

**Comment by doki23** [comment]
> Hi @BohuTANG @doki23
> 
> Could you pls help confirm?
> 
> * Do we support ALL of these aggregate functions for window functions: https://databend.rs/doc/reference/functions/aggregate-functions
> * Do we now support any of non-aggr window functions?
> 
> Thanks.

We support all aggregate functions but not yet implement any non-aggr functions.

**Comment by soyeric128** [comment]
> > Hi @BohuTANG @doki23
> > Could you pls help confirm?
> > 
> > * Do we support ALL of these aggregate functions for window functions: https://databend.rs/doc/reference/functions/aggregate-functions
> > * Do we now support any of non-aggr window functions?
> > 
> > Thanks.
> 
> We support all aggregate functions(any function that impl trait AggregateFunction) but have not yet implemented any non-aggr functions.

Thanks for reply. 
Pls review this PR for doc updates: https://github.com/datafuselabs/databend/pull/6445

**Pull request merged**